src/components/implementation/tests/slm/main.c

```diff
@@ -0,0 +1,405 @@
+#include <cos_component.h>
+#include <llprint.h>
+#include <static_slab.h>
+#include <ps_list.h>
+#include <crt.h>
+
+/***
+ * A version of scheduling using a simple periodic timeout, and fixed
+ * priority, round-robin scheduling.
+ */
+
+#include <slm.h>
+#include <quantum.h>
+#include <fprr.h>
+#include <slm_api.h>
+
+struct crt_comp self;
+
+struct slm_resources_thd {
+	struct crt_thd crt_res;
+};
+
+/*
+ * The thread structure that is a container including
+ *
+ * - core slm thread
+ * - scheduling policy data
+ * - timer policy data
+ * - resources allocated from the kernel via `crt`
+ */
+struct slm_thd_container {
+	struct slm_thd           thd;
+	struct slm_sched_thd     sched;
+	struct slm_timer_thd     timer;
+	struct slm_resources_thd resources;
+};
+
+struct slm_timer_thd *
+slm_thd_timer_policy(struct slm_thd *t)
```

Comment:
Replace with `slm_container.h` (see below).

Create a set of macros to enable the scheduler to easily mix and match different policies (edf/fprr/fifo), timers (quantum/oneshot), and memory allocators (static/dynamic).